README.md
# BattleReporter [![SensioLabsInsight](https://insight.sensiolabs.com/projects/264d2540-4350-4ffb-920c-1967e1db1df3/mini.png)](https://insight.sensiolabs.com/projects/264d2540-4350-4ffb-920c-1967e1db1df3)

Advanced editable BattleReport platform for EVE Online.

- Current Version: `0.4`
- Download at [`Releases`](https://github.com/ta2edchimp/BattleReporter/releases)

## Setup

`BattleReporter` requires an environment running `PHP 5.4`. Version `5.5` is recommended, though.

### Requirements:

- [PhealNG](https://github.com/3rdpartyeve/phealng/)
- [Twig](http://twig.sensiolabs.org/)
- [Slim Framework](http://slimframework.com/)
- [Whoops](https://github.com/filp/whoops)
- [Whoops Middleware for Slim](https://github.com/zeuxisoo/php-slim-whoops)
- [phpFastCache](http://www.phpfastcache.com/)
- [Parsedown](https://github.com/erusev/parsedown)

The dependencies will be met by the use of [Composer](http://getcomposer.org/) during the automatic installation process.

Ensure to have url rewriting enabled and set up properly in your web server for `Slim` to function as expected (see the corresponding [documentation](https://github.com/codeguy/Slim#setup-your-web-server) for more information on how to set this up correctly).

### Installation:

As of release `0.1.6`, a basic installer comes with BattleReporter. Run

	$ php install.php

to launch it. The installer asks for some basic information regarding your installation of BattleReporter, uses [Composer](http://getcomposer.org) to meet the requirements (it will be installed into the installation directory if not present), create the BattleReporter's own database tables and set up the primary admin's user account.

#### Log in via EVE-O SSO

As of release `0.2`, logging into the BattleReport with your EVE Online account is available via Single Sign On. In order to enable log in via EVE-O SSO, you must register your BattleReporter installation as an application in the [EVE Developers portal](https://developers.eveonline.com/). A verified e-mail address is required to register (don't worry this takes no more than hitting a button in the EVE Online Account Management and clicking a link in the e-mail you'll receive soon afterwards).

When registering your BattleReporter installation in the EVE-Dev. portal, ensure to enter the correct `Callback URL`. Let's assume you installed BattleReporter to be available at `br.yourcorp.net` and you do not use SSL, the correct url would be

	http://br.yourcorp.net/login/eve-sso-auth

After hitting `Create Application`, you'll acquire a `Client ID` and a `Secret Key`. These are the values required during BattleReporter's installation process.

#### User Permissions by Role in Corp

As of release `0.3.5`, it is possible to provide an EVE Online API Key to let BattleReporter check its owner corporation's members and their roles. Users logging in with characters who possess the `Director` role can then get the permission to edit, (un)publish and delete other users' battle reports.

For this to work, when setting up an [API Key](https://community.eveonline.com/support/api-key/), it is required to be of type `Corporation` and having access to `MemberSecurity` (AccessMask: `512`).

## Update

As of release `0.3.3`, an update script will assist you in the process of updating BattleReporter's components. Run

	$ php update.php

to launch it. It will update [Composer](http://getcomposer.org), update existing and install new dependencies, as well as update the database tables.

You may use the updater to update or modify the EVE Online Data at anytime. Just replace the files within `/database` and run the update script again. (_Hint:_ the tables required for operation are `mapSolarSystems`, `invGroups` and `invTypes`).

**Important Note when updating any pre `1.0` version:** Please ensure, all configuration variables from `/config.blueprint.php` are mentioned in your `/config.php` after running BattleReporter's update script (For example all variables with the `Slack` default values, when updating from `0.3` to `0.4`).

## Further Configuration

Some of BattleReporter's functions are not yet covered by the `install` and `update` scripts. You'll have to setup and configure them manually by customizing your `config.php` file accordingly.

### Integration to Slack

I sure hope, you know what `Slack` is. If not, head over to the [`Slack` web site](https://slack.com/) and see, why it is better for your corporation's communiation than any IRC, WhatsApp group or whatever you nullsec dwellers currently use to ping the cap pilots ;P

BattleReporter supports to modes to be integrated with `Slack`:

* **Incoming Webhook** configure BattleReporter to automatically post new reports to a certain channel within your Slack team.
* **Slash Command** type `/battlereporter` in any public group or direct conversation to post a BattleReport.

#### Setup as Incoming Webhook

In order for BattleReporter to post anything to `Slack`, you'll have to set up an [Incoming Webhook integration](https://slack.com/services/new/incoming-webhook); Select the `Post to Channel` where you want the reports to be posted and copy the `Webhook URL` provided.

You may also set a label and customize the name and icon, but that's not necessary, as the name as who the BattleReporter posts will be the name of your BattleReporter installation and the icon next to each post will be your corporation's logo.

Next, head over to your `config.php` file and change or add (if yet absent) the following lines:

```php
// Enable Slack Integration
define('BR_API_SLACK_ENABLED', true);
// Set the designated destination for Posts to Slack
define('BR_API_SLACK_CHANNEL', 'insert Webjook URL here');
```

That's it. Everytime you hit `Save and Publish BattleReport` on creation of a new report, it will instantly be posted in the specified channel.

#### Setup Slash Command

You can also set up a Slash Command in `Slack` for BattleReporter. It will enable to directly post BattleReports in the channel you're currently chatting (except: private groups, due to Slack's API limitations). **You still must configure an Incoming Webhook** for the Slash Command to work as expected!

Head over to your Slack Team's settings, and [setup a new Slash Command](https://bitslix.slack.com/services/new/slash-commands):

* Enter `/battlereporter` or whatever you want, into the `Choose a Command` input.
* Enter `http://your.battlereporter.host/api/slack` into the `URL` input.
* Copy the `Token`.

The `Token` is the required information to insert into your BattleReporter installation's `config.php`, change or add:

```php
// Set the authorization Token for Slack Slash Commands
define('BR_API_SLACK_TOKEN', 'insert Token here');
```

You may proceed to configure the `Slash Command` in Slack. Recommended settings are:

* **Activate** `Show this command in the autocomplete list`.
* Set **Description** to sth. like `Post BattleReports`.
* Give **Usage hints**: `[help, list, latest, id]`.

**Usage:**

Type `/battlereporter help` into the current channel's chat to get a list of options; Basically: enter `/battlereporter` (with or without appended `latest`) to post the latest created BattleReport, enter `/battlereporter list` to get a list of the 25 most recently created reports, or use `/battlereporter 123` to post the BattleReport with Id #123.

## Acknowledgements & Legal Stuff

BattleReporter uses parts of [CCPs](http://www.ccpgames.com/) EVE Online SDE, specifically [Steve Ronuken/fuzzysteve's EVE SDE MySQL-port Dump](https://www.fuzzwork.co.uk/dump/latest/).


_to be continued ..._


CHANGELOG
# Release 0.3.3

661120f Version Bump.
9bfa1af Merge branch 'feature/updater-basic' into develop
83677a3 Updated ...
d2dd1f3 Create updater, closes #29
36c1638 Fixed indetation ...
3ded070 Fixes #59
8a1ce2d Merge branch 'feature/proper-logging' into develop
7a0112b Added `DateTimeFileLogger` to enable proper loggin, closes #56
44ffb64 Reworked default `phpFastCache` options to be more consistent.
ea6d0d9 Fixed version property location within json.
26859e9 Remove obsolete variable. Closes #57
63ed2d3 Version Bump.

# Release 0.3.2

469e23b Merge branch 'release/0.3.2'
4bf3395 Merge branch 'feature/db-setup-fix' into develop
96f360b Some further rework of the install process.
f338def Mostly: rework the database imports.
5b63757 Implement import method. This way an upcoming updater does not need to implement this functionality of its own.
8fcf362 Holding pdo object persistent and fixed init command.
eb0e2a4 Changed back to persistent pdo mode.
55d3a10 Changes due to Db handling rework.
434e041 Reworked Db handling.
94e3ed2 aaand undone again, as md skills yet seem to suck. grrr...
dc60c70 Renamed to fully support Markdown on Github.
e88c02a Merge branch 'release/0.3.1' into develop

# Release 0.3.1

d3a91aa Version Bump.
6129354 Merge branch 'release/0.3.0' into develop

# Release 0.3.0

d1f663d Merge branch 'develop' of github.com:ta2edchimp/battlereporter into develop
3077270 Added clickable `footage` icon to "pov-pilots" on BattleReport. Closes #47
d7a8d0f Added clickable `footage` icon to "pov-pilots".
fce3e33 Updated schema for `cyno` feature.
baf4b52 Reincluded detection of already imported kills.
c824fa1 Mark cynos, closes #54
a5787e2 Reworked 3rd-party display on Index.
cc0ef75 Fix #53
bb8cccb Fix #55
4fae4e2 Fix #52

# Release 0.2.8

a10d269 Merge branch 'release/0.2.8' - fixes #36 - fixes #44
6a50251 Version Bump.
18d131e Merge branch 'feature/kb-fetch-newday' into develop Fixes #44
0c8f514 Changed pattern for check of `timespan` user input to work the new way which allows for inputs like `15-2-9 1:23 - 2:34` instead of `2015-02-09 01:23 - 02:34`.
382b2c1 Changed parsing of start and end time from user `timespan` string.
02ecc06 Fix minor parsing error.
07cf3df Merge branch 'feature/br-ship-grouping' into develop
e40ee37 Implement a better way to sort combatants within a battle party.
79a979d Fill BR Ship Groups tables with values for `Logitics`, `Ewar` and `Capitals`.
3862cbe Schema for BR Ship Groups.

# Release 0.2.7

30f7402 Merge branch 'release/0.2.7'
3c3bce2 Merge branch 'hotfix/0.2.6-hotfix-48' into develop
b73368f Merge branch 'hotfix/0.2.6-hotfix-48'
141d0e4 Fixes #48
1ea6a8f Merge branch 'hotfix/v0.2.6-hotfix-46' into develop
5888ed4 Merge branch 'hotfix/v0.2.6-hotfix-46'
f0f0340 Fixes #46
d3b91f3 Merge branch 'master' into develop
83ce5e2 Merge branch 'hotfix/issue-45' into develop
d6edc83 Merge branch 'hotfix/issue-45'
05fdcb8 Fix #45
52b2569 Merge branch 'release/0.2.6' into develop

# Release 0.2.6

f731d34 Merge branch 'release/0.2.6'
06acb17 ID fields are now hidden, as they should be.
5b3de4d Redirect on successfull saving reactivated.
6fbfa9f Updated database schemas.
7d2d7b1 Merge branch 'feature/multiple-pov-footage' into develop
2d4d573 Assign combatants from battle report to connected videos, closes #34
2fd65d9 Reorder added videos.
d12cdc4 Add possibility to add multiple videos to a battle report.

# Release 0.2.5

db2c3f5 Enable direct linking to comments and footage of battle reports, closes #40
2c4b359 Changed response to dismissing the editing of a battle report.
303ba51 Propulate number of pilots per team in `brBattles`, simplifies `Index` query and correctscombatant count, fixes #42
63c25d2 Support YouTube short urls, closes #37
c8c9e7c Test i query returned any value at all, fixes #38
70bb888 Fix #39 corporation / alliance update on EVE SSO login
45c71bc mini bugfix ...
4b2ab4b Changed path concatenation
2331413 Activate caching on productive environments to speed things up a bit.
8421d4e `time_ago` Twig filter was missing the case for < 1d but ≥ 1 day.

# Release 0.2.4

8404aca Use mini `SLInsight` widget.
0be94af Added hint for videos, included proper tooltip for `hasFootage` and `hasComments` icons. Closes #33
ad0ac88 Set `erusev/parsedown` package requirement from `dev-master` to (final) `1.*`.
5c47258 Make use of class `Admin` (where needed), to prevent cluttering the global namespace with admin-functions.
cbf163e Use `Slim::getInstance()` instead of `global`.
d331bd0 Use `Db::getInstance()` instead of `global`.
ed870a8 Register db-accessor as globally accessible instance, to prevent use of `global`.
3d732ef Implemented ability to create a life-cycle dependent, global instance of the db accessor, if needed.
430f698 Code cleanup & prevent `global`.
eca035c Use strict comparison where possible.
12a8f65 Use strict comparison where possible.
3e0a27e Added nl at eof, to comply with code quality policies.
a0e59d8 Throw proper exception instead of the previously used `die()` quirks.
e133756 Now, correctly redirect the `Slim` way ...
5fa3f2b Updated.
a214712 Updated.
426d4a2 Removed.
3603a0b Exclude `.htaccess` from vcs, as this is rather installation specific and should not be distributed.
601fecc Added license field to comply with composer schema.
050ab4e Code cleanup.
5b6ca38 Removed unnecessary call of `exit()`, to not imply unreachable code ...
4e09432 Code cleanup.
e759451 Removed unused variable.
d823144 Removed unused method stub.
3458dec Removed unused variables.
0d6b3d2 Use of type hinting, where possible.
bf21410 Added project name and a discription.
9f2f507 meh ...
47c5a23 Added default `favicon`
f7f563d Updated for removed `DAO.php` include ...
e0a1243 Removed (unused).
ff834f1 Implemented comments, closes #23
54a89ec Included ci medal (if earned sometime).
0295899 Added missing semicolon
6626e1a Meh ...
94064c8 A working errors page, eventually
eda5c21 Redirect to referer if successfully logged in via SSO.

# Release 0.2.3


api/slack.php
<?php

require_once ("$basePath/classes/API/Slack.php");

$app->status(200);
if ($app->request->isPost())
	$app->contentType('application/json');

$params = $app->request->post();
if ($app->request->isGet()) {
	$params = $app->request->get();
}

if (!$app->request->isPost() || BR_API_SLACK_ENABLED !== true || $app->request->post('token') != BR_API_SLACK_TOKEN) {
	// Die silently
	$app->stop();
}


/**

	Init

**/
$command		= "/battlereporter";

$mode			= "latest";
$battleReportID	= -1;

$slackOpts		= array();

if (isset($params["channel_name"])) {
	
	$slackOpts["channel"] = "#" . $params['channel_name'];
	
	if ($slackOpts["channel"] == "#privategroup") {
		echo "_Sorry_, I cannot post to private groups.";
		$app->stop();
	}
	
}


if (isset($params["user_id"]) && !empty($params["user_id"]) && isset($params["user_name"]) && !empty($params["user_name"])) {
	$slackOpts["invUserID"]	= $params["user_id"];
	$slackOpts["invUserName"]	= $params["user_name"];
}


if (isset($params["command"]) && !empty($params["command"])) {
	$command	= $params["command"];
}


/**

	Check current Command and its Argument

**/
if (isset($params['text'])) {
	
	if (is_int($params['text']) || intval($params['text']) > 0) {
		
		$mode			= 'show';
		$battleReportID	= intval($params['text']);
		
	} else {
		
		$mode = strtolower($params['text']);
		if (empty($mode))
			$mode = 'latest';
		
		if ($mode !== 'latest' && $mode !== 'list') {
			
			echo "*" . (BR_OWNER !== '' ? (BR_OWNER . "'s ") : "") . "BattleReporter Help*\n\n" .
				 "Use `$command` without any argument or together with `latest` to post the latest BattleReport.\n\n" .
				 "Use `$command list` to list the last up to 25 BattleReports.\n\n" .
				 "Use `$command 123` to post the BattleReport with ID `123`.\n\n" .
				 "*BattleReporter cannot post to private groups!*";
			
			$app->stop();
			
		}
		
	}
	
}


/**

	List the last (max.: 25) published BattleReports

**/
if ($mode === 'list') {
	
	$result = Battle::getList(array(
		"onlyPublished" => true,
		"count" => 25
	));
	
	if ($result === NULL || $result === FALSE) {
		echo "_Sorry_, something went wrong.";
		$app->stop();
	}
	
	if (count($result) < 1) {
		echo "_Sorry_, but there aren't any published BattleReports, yet.";
		$app->stop();
	}
	
	$sampleID		= 0;
	$sampleTitle	= "";
	
	foreach ($result as $battle) {
		
		if ($sampleID == 0) {
			$sampleID = $battle["battleReportID"];
			$sampleTitle = (empty($battle["title"]) ? ("Battle in " . $battle["solarSystemName"]) : $battle["title"]);
		}
		
		echo "*" . $battle["battleReportID"] . "* - " .
			 (empty($battle["title"]) ? ("*Battle in " . $battle["solarSystemName"] . "*\n") : ("*" . $battle["title"] . "*\nin " . $battle["solarSystemName"])) . 
			 " on " . date("Y-m-d H:i", $battle["startTime"]) .
			 "\n\n";
		
	}
	
	if ($sampleID == 0) {
		echo "*Hint:*\nType `$command 123` to post BattleReport #123 (example).";
	} else {
		echo "*Hint:*\nType `$command $sampleID` to post BattleReport #$sampleID\n*$sampleTitle*.";
	}
	
	
	$app->stop();
	
}


/**

	Show the latest or another specified BattleReport

**/
$battleToShow = null;

if ($mode === 'latest') {
	
	$result = Battle::getList(array(
		"onlyPublished" => true,
		"count" => 1
	));
	
	if ($result === NULL || $result === FALSE || count($result) !== 1) {
		echo "_Sorry_, something went wrong.";
		$app->stop();
	}
	
	$battleToShow = $result[0];
	
	$mode			= 'show';
	$battleReportID	= $battleToShow["battleReportID"];
	
}

if ($mode !== 'show') {
	echo "*Unknown command `$mode`*.";
	$app->stop();
}

$slack = new Slack(BR_API_SLACK_CHANNEL);
$battleToShow = $slack->postBattleWithID($battleReportID, $slackOpts);

if ($battleToShow === null) {
	echo "*Invalid or unknown BattleReport ID* `$battleReportID`.";
	$app->stop();
}


/**

	Send Response to user

**/
echo "Posted Battle" .
	 (!empty($battleToShow["title"]) ? (" \"" . $battleToShow["title"] . "\"") : (" in " . $battleToShow["solarSystemName"])) .
	 " to Slack" .
	 (isset($slackOpts["channel"]) ? (", channel `" . $slackOpts["channel"] . "`") : "") .
	 ".";
